### PR TITLE
init commit

### PR DESCRIPTION
**Pull request by HarlamovBuldog**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package main
+
+import (
+	"./wget"
+)
+
+func main() {
+	wget.Execute()
```
Comment (pisarukv) [inline main.go]: better put all inside main

File: wget/Example_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	for _, test := range tests {
+		link := test.inputLink
+		got, err := WasFileDownloaded(link, dm.fileList[link].totalSize)
+		if got != test.want {
+			t.Errorf("Download(%q) = %v : %v", test.inputLink, got, err.Error())
+		}
+	}
+}
+
+func WasFileDownloaded(link string, totalSizeURL int64) (bool, error) {
```
Comment (pisarukv) [inline wget/Example_test.go]: why a u exporting the function?

File: wget/wget.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+// we are going to download
+type File struct {
+	io.Reader       // To implement Reader interface
+	totalSize int64 // Total content size in bytes
+	counter   int64 // Total # of bytes transferred
+}
+
+type DownloadManager struct {
+	fileList        map[string]*File // key = link
+	fileListOrder   []string         // to order fileList map iteration
```
Comment (pisarukv) [inline wget/wget.go]: fileList can be slice, and you can do it without fileListOrder

File: wget/wget.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+// NewDownloadManager is construct function
+func NewDownloadManager() *DownloadManager {
+	return &DownloadManager{
+		fileList:        make(map[string]*File),
+		fileListOrder:   make([]string, 0),
+		printFinishedCh: make(chan struct{}),
+	}
+}
+
+func ParseURL(link string) error {
```
Comment (pisarukv) [inline wget/wget.go]: What actually dose this function dose? ParseURL but why it returns only error? Need better naming here

File: wget/wget.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	if err != nil {
+		return fmt.Errorf("wget: error connecting url :%v", err)
+		//fmt.Fprintf(os.Stderr, "wget: error connecting url :%v\n", err)
+	}
+	defer resp.Body.Close()
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if err != nil {
+		return fmt.Errorf("wget: content length = %v :%v", bodySize, err)
+		//fmt.Fprintf(os.Stderr, "wget: error getting content length :%v\n", err)
```
Comment (pisarukv) [inline wget/wget.go]: delete dead code always

File: wget/wget.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	}
+	return nil
+}
+
+func Execute() {
+	wg := &sync.WaitGroup{}
+	var linksList []string
+	dm := NewDownloadManager()
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (pisarukv) [inline wget/wget.go]: ctx, cancel := context.WithCancel(context.Background())

File: wget/wget.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+		return fmt.Errorf("wget: content length = %v :%v", bodySize, err)
+		//fmt.Fprintf(os.Stderr, "wget: error getting content length :%v\n", err)
+	}
+	return nil
+}
+
+func Execute() {
+	wg := &sync.WaitGroup{}
+	var linksList []string
+	dm := NewDownloadManager()
```
Comment (pisarukv) [inline wget/wget.go]: create vars just before the place where are they used

File: wget/wget.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+	for _, inputLink := range os.Args[1:] {
+		err := ParseURL(inputLink)
+		if err == nil {
+			linksList = append(linksList, inputLink)
+		} else {
+			fmt.Fprintf(os.Stderr, "wget: error parsing url: %v\n", err)
+		}
+	}
+
+	// locking map here
```
Comment (pisarukv) [inline wget/wget.go]: dm := NewDownloadManager()

File: wget/wget.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+// print order of map fields during iteration
+func (dm *DownloadManager) Print(ctx context.Context) {
+	defer close(dm.printFinishedCh)
+
+	print := func() {
+		fmt.Print("\r")
+		dm.fileListLock.Lock()
+		for _, key := range dm.fileListOrder {
+			if dm.fileList[key].totalSize != 0 {
+				fmt.Printf("%.2f %% ", float64(dm.fileList[key].counter)/float64(dm.fileList[key].totalSize)*100)
```
Comment (pisarukv) [inline wget/wget.go]: put into a separate var

File: wget/wget.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		dm.fileListLock.Unlock()
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			print()
+			return
+		default:
```
Comment (cali4888) [inline wget/wget.go]: it's possible here to write case <- time.NewTicker(100 * time.Millisecond):

File: wget/wget.go [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+	if err != nil {
+		//return false, 0, fmt.Errorf("wget: error connecting url :%v", err)
+		fmt.Fprintf(os.Stderr, "wget: error connecting url :%v\n", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if err != nil {
+		//return false, 0, fmt.Errorf("wget: error getting content length :%v", err)
```
Comment (pisarukv) [inline wget/wget.go]: dead code

File: wget/wget.go [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+func (dm *DownloadManager) WaitPrintEnd() {
+	<-dm.printFinishedCh
+	return
+}
+
+func (dm *DownloadManager) Download(link string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := http.Get(link)
+	if err != nil {
+		//return false, 0, fmt.Errorf("wget: error connecting url :%v", err)
```
Comment (pisarukv) [inline wget/wget.go]: dead code

File: wget/wget.go [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+		return
+	}
+
+	if err := f.Close(); err != nil {
+		//return false, int(bytes), fmt.Errorf("wget: error closing file %s: %v", f.Name(), err)
+		fmt.Fprintf(os.Stderr, "wget: error closing file %s: %v\n", f.Name(), err)
+		return
+	}
+
+	//return true, int(bytes), nil
```
Comment (pisarukv) [inline wget/wget.go]: dead code

File: wget/wget.go [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+	dm.fileListLock.Lock()
+	file := dm.fileList[link]
+	file.Reader = resp.Body
+	file.totalSize = bodySize
+	dm.fileListLock.Unlock()
+
+	// Copy need to be changed to Tee.Reader
+	_, err1 := io.Copy(f, file)
+	if err1 != nil {
+		//return false, 0, fmt.Errorf("wget: error reading from %s: %v", link, err)
```
Comment (pisarukv) [inline wget/wget.go]: dead code

File: wget/Example_test.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
+package wget
```
Comment (pisarukv) [inline wget/Example_test.go]: Example_test.go -> wget_test.go

File: main.go [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package main
+
+import (
+	"./wget"
```
Comment (pisarukv) [inline main.go]: in go it's common to use absolut imports

Comment (pisarukv) [reply main.go]: so in your case it would be `gitHub.com/HarlamovBuldog/wget-golang-util/wget`

File: wget/Example_test.go (hunk 2 above)
Comment (pisarukv) [reply wget/Example_test.go]: And of course isFileDownloaded, but more go way it to name it fileDownloaded

Comment (pisarukv) [reply wget/Example_test.go]: pls check https://golang.org/doc/effective_go.html

File: wget/wget.go (hunk 5 above)
Comment (HarlamovBuldog) [reply wget/wget.go]: Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

File: wget/wget.go (hunk 11 above)
Comment (HarlamovBuldog) [reply wget/wget.go]: Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

File: wget/wget.go (hunk 12 above)
Comment (HarlamovBuldog) [reply wget/wget.go]: Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

File: wget/wget.go (hunk 13 above)
Comment (HarlamovBuldog) [reply wget/wget.go]: Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

File: main.go [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+			print()
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+}
+
+// Download tries to download file from remote address using
+// given url. Takes WaitGroup pointer to provide parallelism.
+// Prints error to os.Stderr and return if something goes wrong.
+func (file *File) Download(link string, wg *sync.WaitGroup) {
```
Comment (pisarukv) [inline main.go]: do you want download to be public method?

File: main.go [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+	for _, inputLink := range os.Args[1:] {
+		parsedLink, err := parseURL(inputLink)
+		if err == nil {
+			linksList = append(linksList, parsedLink)
+		} else {
+			fmt.Fprintf(os.Stderr, "wget: error parsing url %v: %v\n", inputLink, err)
+		}
+	}
+
+	var fileList []*File
```
Comment (pisarukv) [inline main.go]: at this moment you already know number of links, so why not fileList := make([]*File, Len(linksList)) ?

File: main.go [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+	// creating WaitGroup for download goroutines
+	wg := &sync.WaitGroup{}
+
+	for _, link := range linksList {
+		file := &File{}
+		fileList = append(fileList, file)
```
Comment (pisarukv) [inline main.go]: fileList = append(fileList, &File{})

File: main.go [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+			if file.totalSize != 0 {
+				percentage := float64(file.counter) / float64(file.totalSize) * 100
+				fmt.Printf("%.2f %% ", percentage)
+			} else {
+				fmt.Print("0.00 %% ")
+			}
+		}
+	}
+
+	for {
```
Comment (pisarukv) [inline main.go:122]:
```
	for {
 		select {
 		case <-ctx.Done():
 			print()
 			return
 		default:
 			print()
 			time.Sleep(100 * time.Millisecond)
 		}
 	}
```

```
        t := time.NewTicker(100 * time.Millisecond)
        defer t.Stop()
	for {
 		select {
 		case <-ctx.Done():
 			print()
 			return
 		case <- t.C:
 			print()
 		}
 	}
```

File: main.go [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+func (file *File) Download(link string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := http.Get(link)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "wget: error connecting url :%v\n", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
```
Comment (pisarukv) [inline main.go]: so, you already getting and checking content-length in parse url? why you need to do double work?

File: main.go [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "wget: error getting content length :%v\n", err)
```
Comment (pisarukv) [inline main.go]: log.Printf - will write err to stderr

File: main.go [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+		fmt.Fprintf(os.Stderr, "wget: error creating file :%v\n", err)
+		return
+	}
+
+	file.name = fileNameToSave
+	file.Reader = resp.Body
+	file.totalSize = bodySize
+
+	// Copy need to be changed to Tee.Reader
+	_, err1 := io.Copy(f, file)
```
Comment (pisarukv) [inline main.go]: why err1?

Comment (HarlamovBuldog) [reply main.go]: cause err was declared below. If I will name it err, vscode will tell me smth like "No new variable was declared on the left side"

Comment (HarlamovBuldog) [reply main.go]: forget, i can just type = instead of :=

File: main.go (hunk 19 above)
Comment (HarlamovBuldog) [reply main.go]: But how then i will run file.Download method after append? I mean i need this pointer.

Comment (pisarukv) [reply main.go]: agree

**Review by pisarukv — APPROVED** [review]
looks good

**Pull request merged**